### PR TITLE
Fix cyclic record resolution

### LINKED ISSUE #30347: The compilation of `OAuth2` module build is failing intermittently
**Description:**
The compilation of [OAuth2 module](https://github.com/ballerina-platform/module-ballerina-oauth2) build is failing intermittently with the lang version `Ballerina Lang : 2.0.0-beta.1-20210504-152000-4d5f126d`. This happens in both Windows and Ubuntu builds.

```
Compiling source
	ballerina/oauth2:1.1.0-beta.1
ERROR [listener_oauth2_provider.bal:(105:9,105:19)] unknown type 'ClientAuth'
ERROR [listener_oauth2_provider.bal:(105:28,105:66)] invalid operation: type 'ballerina/oauth2:1.1.0-beta.1:ClientConfiguration' does not support optional field access for field 'auth'
ERROR [listener_oauth2_provider.bal:(106:21,106:31)] unknown type 'ClientAuth'

ERROR [oauth2_commons.bal:(31:5,31:15)] unknown type 'ClientAuth'
> Task :oauth2-ballerina:ballerinaBuild FAILED
error: compilation contains errors
```

The type which says as unknown type is defined as `public type ClientAuth ClientCredentialsGrantConfig|PasswordGrantConfig|RefreshTokenGrantConfig;` It is a union of records.

This is the PR, that the build is failing. 1 build has failed out of 4. That can be changed when retriggering, since this is intermittent.
https://github.com/ballerina-platform/module-ballerina-oauth2/pull/172

**Affected Versions:**
Ballerina Lang : 2.0.0-beta.1-20210504-152000-4d5f126d

### LINKED ISSUE #30356: Type resolution order does not consider fields
**Description:**

Consider the below ballerina code. Here by the time **record-type-inclusion** check whether `Baz` is anydata type we have not resolved its fields. At this moment This is just an empty shell. In this situation type checker says `Baz` is not an anydata type. This leads to the redeclared symbol error message.

```Ballerina
public type Foo record {|
    anydata body;
|};

// type Baz record {|
//     int id;
// |};

type Bar record {| 
    *Foo;
    Baz body;
|};

type Baz record {|
    int id;
|};
```

- Tried to fix this by adding a type definition to the unresolved list if there is an unresolved field also in type resolution logic. But this started causing cyclic record via fields to be resolved intermittently causing issues. PR https://github.com/ballerina-platform/ballerina-lang/pull/29987
- Intermittent fail recognized in **OAuth2** module. https://github.com/ballerina-platform/ballerina-lang/issues/30347
- Code is reverted by PR: https://github.com/ballerina-platform/ballerina-lang/pull/30348

To fix this problem completely below cases needs to be handled
- [ ] Fix type checker issue
- [ ] Enable disabled test cases by PR 30348
- [ ] Verify fix with all the new scenarios 

#### Single bal files

<details>
<summary>
<a class="btnfire small stroke"><em class="fas fa-chevron-circle-down"></em>&nbsp;&nbsp;Case 1 - code</a>    
</summary>
<p>

```Ballerina
public type Foo record {|
    anydata body;
|};

type Bar record {| 
    *Foo;
    Baz body1;
|};

type Baz record {|
    int id;
|};
```

</p>
</details>

<details>
<summary>
<a class="btnfire small stroke"><em class="fas fa-chevron-circle-down"></em>&nbsp;&nbsp;Case 2- code</a>    
</summary>
<p>

```Ballerina
public type ClientCredentialsGrantConfig record {|
    string tokenUrl;
    ClientConfiguration clientConfig = {};
|};

type AnydataConfig record {|
    anydata auth?;
|};

public type ClientConfiguration record {|
    *AnydataConfig;
    ClientAuth auth?;
|};

# Description  
public type ClientAuth ClientCredentialsGrantConfig;

public function func(ClientAuth cl) {
    anydata a = cl; // OK
}
```

</p>
</details>

#### Projects
TODO: add more

<details>
<summary>
<a class="btnfire small stroke"><em class="fas fa-chevron-circle-down"></em>&nbsp;&nbsp;Case 1- code</a>    
</summary>
<p>

proj/client_oauth2_provider.bal
```Ballerina
public type ClientCredentialsGrantConfig record {|
    string tokenUrl;
    ClientConfiguration clientConfig = {};
|};
proj/oauth2_commons.bal

public type ClientConfiguration record {|
    ClientAuth auth?;
|};
```

proj/client_oauth2_provider.bal
```Ballerina
public type ClientCredentialsGrantConfig record {|
    string tokenUrl;
    ClientConfiguration clientConfig = {};
|};
proj/oauth2_commons.bal

public type ClientConfiguration record {|
    ClientAuth auth?;
|};

public type ClientAuth ClientCredentialsGrantConfig;
```

</p>
</details>

### PR DESCRIPTION
**Pull request by KRVPerera**
## Purpose

Fixes #30356

## Remarks
Related to #30347

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/record/cyclic_record_via_fields.bal [hunk 1]
```diff
@@ -0,0 +1,9 @@
+public type ClientConfiguration record {|
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/record/cyclic_record_via_fields.bal:1]: Shall we add a project with the two files like in https://github.com/ballerina-platform/ballerina-lang/issues/30347#issuecomment-832030912 and test instead? I was unable to reproduce this issue with a single bal file.

Also missing the licence header.

Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/resources/test-src/record/cyclic_record_via_fields.bal:1]: Sure, I will add. I was able to reproduce the issue with this perticular test case.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/record/cyclic_record_via_fields.bal:1]: Oh okay, if you were able to reproduce it with this let's keep this test also.

**Comment by MaryamZi** [comment]
The following sample works on alpha5, but doesn't now.

```ballerina
public type ClientCredentialsGrantConfig record {|
    string tokenUrl;
    ClientConfiguration clientConfig = {};
|};

type AnydataConfig record {|
    anydata auth?;
|};

public type ClientConfiguration record {|
    *AnydataConfig;
    ClientAuth auth?;
|};

public type ClientAuth ClientCredentialsGrantConfig;

public function func(ClientAuth cl) {
    anydata a = cl; // OK
}
```

```cmd
ERROR [foo.bal:(11:6,11:19)] redeclared symbol 'auth'
ERROR [foo.bal:(18:17,18:19)] incompatible types: expected 'anydata', found 'ClientCredentialsGrantConfig'
```

The second error is not observed when this is part of a project.
proj/client_oauth2_provider.bal
```ballerina
public type ClientCredentialsGrantConfig record {|
    string tokenUrl;
    ClientConfiguration clientConfig = {};
|};
```

proj/oauth2_commons.bal
```ballerina
type AnydataConfig record {|
    anydata auth?;
|};

public type ClientConfiguration record {|
    *AnydataConfig;
    ClientAuth auth?;
|};

public type ClientAuth ClientCredentialsGrantConfig;

public function func(ClientAuth cl) {
    anydata a = cl; // OK
}
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 2]
```diff
@@ -713,6 +712,8 @@ private void defineReferencedClassFields(BLangClassDefinition classDefinition, S
                     if (types.isAssignable(existingVariable.type, field.type)) {
                         continue;
                     }
+                    dlog.error(existingVariable.pos, DiagnosticErrorCode.INCOMPATIBLE_SUB_TYPE_FIELD,
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:715]: Is this not covered by tests? There seems to be a codecov comment.

**Comment by MaryamZi** [comment]
@KRVPerera, this seems to have a conflict too.

**Comment by MaryamZi** [comment]
@KRVPerera, org.ballerinalang.test.expressions.rawtemplate.RawTemplateLiteralTest.testNegatives seems to be failing.

```
incorrect error message: expected [redeclared symbol 'strings'] but found [included field 'strings' of type '(string[] & readonly)' cannot be overridden by a field of type 'string[]': expected a subtype of '(string[] & readonly)']
```

Shall we check and update the PR?

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**